lib/nrf9160_time/nrf9160_time.c

```diff
@@ -0,0 +1,284 @@
+#include <sys/timeutil.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(nrf9160_time, CONFIG_NRF9160_TIME_LOG_LEVEL);
+
+#define AT_CMD_MODEM_DATE_TIME			"AT+CCLK?"
+#define AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN	28
+
+#define UIO_IP      "129.240.2.6"
+#define GOOGLE_IP   "216.239.35.0"
+#define GOOGLE_IP_2 "216.239.35.4"
+#define GOOGLE_IP_3 "216.239.35.8"
+#define GOOGLE_IP_4 "216.239.35.12"
+
+K_SEM_DEFINE(fetch_time_sem, 0, 1);
+
+static struct k_delayed_work time_work;
+
+static struct time_aux {
+	s64_t date_time_utc;
+	int last_date_time_update;
+} time_aux;
+
+static struct sntp_time sntp_time;
+static bool initial_valid_time;
+
+struct ntp_servers {
+	const char *server;
+};
+
+struct ntp_servers servers[] = {
+	{.server = UIO_IP},
+	{.server = GOOGLE_IP},
+	{.server = GOOGLE_IP_2},
+	{.server = GOOGLE_IP_3},
+	{.server = GOOGLE_IP_4}
+};
+
+#if defined(CONFIG_NRF9160_TIME_MODEM)
+static int parse_time_entries(char *datetime_string, int min, int max)
+{
+	char buf[50];
+
+	for (int i = min; i < max + 1; i++) {
+		buf[i - min] = datetime_string[i];
+	}
+
+	return strtol(buf, NULL, 10);
+}
+
+static int get_time_cellular_network(void)
+{
+	int err;
+	char buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN + 5];
+	struct tm date_time;
+
+	err = at_cmd_write(AT_CMD_MODEM_DATE_TIME, buf, sizeof(buf), NULL);
+	if (err) {
+		LOG_DBG("Could not get cellular network time, error: %d", err);
+		return err;
+	}
+	buf[AT_CMD_MODEM_DATE_TIME_RESPONSE_LEN] = 0;
+
+	LOG_DBG("Response from modem: %s", log_strdup(buf));
+
+	date_time.tm_year = parse_time_entries(buf, 8, 9) + 2000 - 1900;
+	date_time.tm_mon  = parse_time_entries(buf, 11, 12) - 1;
+	date_time.tm_mday = parse_time_entries(buf, 14, 15);
+	date_time.tm_hour = parse_time_entries(buf, 17, 18);
+	date_time.tm_min  = parse_time_entries(buf, 20, 21);
+	date_time.tm_sec  = parse_time_entries(buf, 23, 24);
+
+	if (date_time.tm_year == 115 || date_time.tm_year == 0) {
+		LOG_DBG("Modem giving old cellular network time");
+		return -ENODATA;
+	}
+
+	time_aux.date_time_utc = (s64_t)timeutil_timegm64(&date_time) * 1000;
+	time_aux.last_date_time_update = k_uptime_get();
+
+	return 0;
+}
+#endif
+
+static int sntp_request_time(const char *server, u32_t timeout,
+			     struct sntp_time *time)
+{
+	int err;
+	static struct addrinfo hints;
+	struct addrinfo *addr;
+	struct sntp_ctx sntp_ctx;
+
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_DGRAM;
+	hints.ai_protocol = 0;
+
+	err = net_getaddrinfo_addr_str(server, "123", &hints, &addr);
+	if (err) {
+		LOG_ERR("net_getaddrinfo_addr_str, error: %d", err);
+		return err;
+	}
+
+	err = sntp_init(&sntp_ctx, addr->ai_addr, addr->ai_addrlen);
+	if (err) {
+		goto freeaddr;
+	}
+
+	err = sntp_query(&sntp_ctx, timeout, time);
+	if (err) {
+		sntp_close(&sntp_ctx);
+		goto freeaddr;
+	}
+
+freeaddr:
+	freeaddrinfo(addr);
+	return err;
+}
+
+static int get_time_NTP_server(void)
+{
+	int err;
+	int i = 0;
+
+	while (i < ARRAY_SIZE(servers)) {
```

Comment:
IMHO it's clearer that we are looping over the servers if we use a normal for loop here.